aggregator/src/aggregator/accumulator.rs

```diff
@@ -137,7 +137,22 @@ impl<const SEED_SIZE: usize, Q: AccumulableQueryType, A: vdaf::Aggregator<SEED_S
     ) -> Result<HashSet<ReportId>, datastore::Error> {
         let unmergeable_report_ids = Arc::new(Mutex::new(HashSet::new()));
 
-        try_join_all(self.aggregations.values().map(|data| {
+        // Prevent deadlocks when inserting into shards by providing a deterministic order to shard
+        // updates. Suppose two concurrent processes are attempting to update a batch aggregation.
+        // We want to avert this situation:
+        //
+        // A> BEGIN;
+        // A> UPDATE batch_aggregations WHERE ord = 1 ... -- Row with ord 1 is locked for update.
+        // B> BEGIN;
+        // B> UPDATE batch_aggregations WHERE ord = 2 ... -- Row with ord 2 is locked for update.
+        // A> UPDATE batch_aggregations WHERE ord = 2 ... -- A is now blocked waiting for B to finish.
+        // B> UPDATE batch_aggregations WHERE ord = 1 ... -- Kaboom!
+        //
+        // This situation cannot happen if the UPDATE statements are ordered by `ord`.
+        let mut aggregations: Vec<_> = self.aggregations.values().collect();
+        aggregations.sort_unstable_by_key(|data| data.batch_aggregation.ord());
```

Comment:
nit: a single `Accumulator` instance handles a single `task_id` & `aggregation_param`, but in a time-interval query we might be accumulating into multiple batches at once. I think we should sort on a key that is `(batch_identifier, ord)` rather than just `ord`.

Reply:
Nice catch!

Reply:
This was a bit trickier than I had originally thought, since `QueryType::BatchIdentifier` is not `Ord`.

I don't think it makes sense to make it `Ord`. That'll be a breaking change in the `messages` crate. I also don't think there's a meaningful ordering for `TimeInterval` (indeed even `Range` is not `Ord`).

So I ended up implementing a `compare_batch_identifiers()` function on `AccumulableQueryType` with the noted caveat that it's really only for database ordering, which fits the purpose of the `aggregator_core` crate anyways.

I could use a peek at it since it feels somewhat hacky, maybe there's an easier way.

Reply:
tl;dr: I think we should make `QueryType::BatchIdentifier` and `QueryType::PartialBatchIdentifier` `Ord` (but let's chat about why/how this is a breaking change). Let's make them `PartialOrd` too, for completeness.

> I don't think it makes sense to make it Ord. That'll be a breaking change in the messages crate.

Why is this breaking -- is it only because someone outside of Janus might have implemented the `QueryType` trait (and would now have to make sure their batch identifiers are `Ord`)? If so, IMO we should take this change without considering a major-version bump; I think `QueryType` is intended for Janus usage only, but is defined in `janus_messages` as a public type because (a) `janus_messages` needs to understand query types and (b) dependent application crates also need to understand query types. I don't think we really expect anyone to implement a new query type, and if they do, they making their batch identifier `Ord` seems like a small change even if technically breaking -- batch identifiers ought to be plain-old-datatypes that can be compared lexicographically, if no other comparison is feasible.

> I also don't think there's a meaningful ordering for Interval (indeed even Range is not Ord).

Any consistent definition of `Ord` will work here. For example, deriving `PartialOrd` and `Ord` on `Interval` gives a "lexicographic" comparison, i.e. start time is compared first, then duration is compared if the start times are equal.

Reply:
Lexicographic `Ord` SGTM!

> is it only because someone outside of Janus might have implemented the QueryType trait (and would now have to make sure their batch identifiers are Ord)?

That was my thought, yes.